lib/AST/Type.cpp

```diff
@@ -54,6 +54,12 @@ using namespace swift;
 #include "swift/AST/TypeNodes.def"
 
 Type QueryTypeSubstitutionMap::operator()(SubstitutableType *type) const {
+  if (auto *archetype = dyn_cast<ArchetypeType>(type)) {
```

Comment:
Let's remove this part then, or change it to an assertion.

Type::subst() passing in non-root archetypes is something that I believe only associated type inference relies on. You can try taking it out and see what breaks.

Reply:
I still need this tweak though. Otherwise `QueryTypeSubstitutionMapOrIdentity` won't work with non-root opened archetypes, and that makes the SIL cloner crash. I can use a custom substitution fn instead, WDYT?

Reply:
Would it not be safer to leave this part as is and add a `FIXME`, rather than propagating the check to clients by replacing the function objects with custom lambdas?